File: compiler/extensions/cpp/runtime/test/zserio/BitBufferTest.cpp
#include <utility>

#include "zserio/CppRuntimeException.h"
#include "zserio/BitBuffer.h"

#include "gtest/gtest.h"

namespace zserio
{

TEST(BitBufferTest, emptyConstructor)
{
    const BitBuffer bitBuffer;
    ASSERT_EQ(0, bitBuffer.getBitSize());
    ASSERT_EQ(0, bitBuffer.getByteSize());
}

TEST(BitBufferTest, bitSizeConstructor)
{
    const size_t bitSize = 11;
    const BitBuffer bitBuffer(bitSize);
    ASSERT_EQ(bitSize, bitBuffer.getBitSize());
}

TEST(BitBufferTest, vectorConstructor)
{
    const size_t byteSize = 2;
    const std::vector<uint8_t> buffer(byteSize);
    const BitBuffer bitBuffer(buffer);
    ASSERT_EQ(8 * byteSize, bitBuffer.getBitSize());

    const size_t emptyBitSize = 0;
    const std::vector<uint8_t> emptyBuffer;
    const BitBuffer emptyBitBuffer(emptyBuffer);
    ASSERT_EQ(emptyBitSize, emptyBitBuffer.getBitSize());
}

TEST(BitBufferTest, vectorBitSizeConstructor)
{
    const size_t bitSize = 11;
    const std::vector<uint8_t> buffer((bitSize + 7) / 8);
    const BitBuffer bitBuffer(buffer, bitSize);
    ASSERT_EQ(bitSize, bitBuffer.getBitSize());

    const size_t emptyBitSize = 0;
    const std::vector<uint8_t> emptyBuffer;
    const BitBuffer emptyBitBuffer(emptyBuffer, emptyBitSize);
    ASSERT_EQ(emptyBitSize, emptyBitBuffer.getBitSize());

    const size_t outOfRangeBitSize = 9;
    const std::vector<uint8_t> outOfRangeBuffer(1);
    ASSERT_THROW(BitBuffer(outOfRangeBuffer, outOfRangeBitSize), CppRuntimeException);
}

TEST(BitBufferTest, vectorMoveConstructor)
{
    const size_t byteSize = 2;
    std::vector<uint8_t> buffer(byteSize);
    const uint8_t* bufferStart = buffer.data();
    const BitBuffer bitBuffer(std::move(buffer));
    ASSERT_EQ(8 * byteSize, bitBuffer.getBitSize());
    ASSERT_EQ(bufferStart, bitBuffer.getBuffer());
}

TEST(BitBufferTest, vectorBitSizeMoveConstructor)
{
    const size_t bitSize = 11;
    std::vector<uint8_t> buffer((bitSize + 7) / 8);
    const uint8_t* bufferStart = buffer.data();
    const BitBuffer bitBuffer(std::move(buffer), bitSize);
    ASSERT_EQ(bitSize, bitBuffer.getBitSize());
    ASSERT_EQ(bufferStart, bitBuffer.getBuffer());

    const size_t outOfRangeBitSize = 9;
    ASSERT_THROW(BitBuffer(std::vector<uint8_t>(1), outOfRangeBitSize), CppRuntimeException);
}

TEST(BitBufferTest, rawPointerConstructor)
{
    const size_t bitSize = 11;
    const std::vector<uint8_t> buffer((bitSize + 7) / 8);
    const BitBuffer bitBuffer(buffer.data(), bitSize);
    ASSERT_EQ(bitSize, bitBuffer.getBitSize());

    const size_t emptyBitSize = 0;
    const std::vector<uint8_t> emptyBuffer;
    const BitBuffer emptyBitBuffer(emptyBuffer.data(), emptyBitSize);
    ASSERT_EQ(emptyBitSize, emptyBitBuffer.getBitSize());
}

TEST(BitBufferTest, copyConstructor)
{
    const size_t bitSize = 11;
    const std::vector<uint8_t> buffer = {0xAB, 0xE0};
    const BitBuffer bitBuffer(buffer, bitSize);

    const BitBuffer copiedBitBuffer(bitBuffer);
    ASSERT_EQ(bitBuffer.getBitSize(), copiedBitBuffer.getBitSize());
    ASSERT_EQ(bitBuffer.getByteSize(), copiedBitBuffer.getByteSize());
    const uint8_t* copiedBuffer = copiedBitBuffer.getBuffer();
    for (uint8_t element : buffer)
    {
        ASSERT_EQ(element, *copiedBuffer);
        copiedBuffer++;
    }
}

TEST(BitBufferTest, assignmentOperator)
{
    const size_t bitSize = 11;
    const std::vector<uint8_t> buffer = {0xAB, 0xE0};
    const BitBuffer bitBuffer(buffer, bitSize);

    const BitBuffer copiedBitBuffer = bitBuffer;
    ASSERT_EQ(bitBuffer.getBitSize(), copiedBitBuffer.getBitSize());
    ASSERT_EQ(bitBuffer.getByteSize(), copiedBitBuffer.getByteSize());
    const uint8_t* copiedBuffer = copiedBitBuffer.getBuffer();
    for (uint8_t element : buffer)
    {
        ASSERT_EQ(element, *copiedBuffer);
        copiedBuffer++;
    }
}

TEST(BitBufferTest, moveConstructor)
{
    const size_t bitSize = 11;
    const std::vector<uint8_t> buffer = {0xAB, 0xE0};
    const size_t byteSize = buffer.size();
    BitBuffer bitBuffer(buffer, bitSize);
    const uint8_t* bufferStart = bitBuffer.getBuffer();

    const BitBuffer movedBitBuffer(std::move(bitBuffer));
    ASSERT_EQ(bitSize, movedBitBuffer.getBitSize());
    ASSERT_EQ(byteSize, movedBitBuffer.getByteSize());
    ASSERT_EQ(bufferStart, movedBitBuffer.getBuffer());
}

TEST(BitBufferTest, moveAssignmentOperator)
{
    const size_t bitSize = 11;
    const std::vector<uint8_t> buffer = {0xAB, 0xE0};
    const size_t byteSize = buffer.size();
    BitBuffer bitBuffer(buffer, bitSize);
    const uint8_t* bufferStart = bitBuffer.getBuffer();

    const BitBuffer movedBitBuffer = std::move(bitBuffer);
    ASSERT_EQ(bitSize, movedBitBuffer.getBitSize());
    ASSERT_EQ(byteSize, movedBitBuffer.getByteSize());
    ASSERT_EQ(bufferStart, movedBitBuffer.getBuffer());
}

TEST(BitBufferTest, equalOperator)
{
    const size_t bitSize = 11;
    const BitBuffer bitBuffer1(std::vector<uint8_t>({0xAB, 0xE0}), bitSize);
    const BitBuffer bitBuffer2(std::vector<uint8_t>({0xAB, 0xF0}), bitSize);
    ASSERT_TRUE(bitBuffer1 == bitBuffer2);

    const BitBuffer bitBuffer3(std::vector<uint8_t>({0xAB, 0xFF}), bitSize);
    ASSERT_TRUE(bitBuffer1 == bitBuffer3);

    const BitBuffer bitBuffer4(std::vector<uint8_t>({0xAB, 0xC0}), bitSize);
    ASSERT_FALSE(bitBuffer1 == bitBuffer4);

    const BitBuffer bitBuffer5(std::vector<uint8_t>({0xBA, 0xE0}), bitSize);
    ASSERT_FALSE(bitBuffer1 == bitBuffer5);

    const BitBuffer bitBuffer6(std::vector<uint8_t>({0xAB}));
    ASSERT_FALSE(bitBuffer1 == bitBuffer6);

    const BitBuffer bitBuffer7;
    ASSERT_FALSE(bitBuffer1 == bitBuffer7);
}

TEST(BitBufferTest, hashCode)
{
    const size_t bitSize = 11;
    const BitBuffer bitBuffer1(std::vector<uint8_t>({0xAB, 0xE0}), bitSize);
    const BitBuffer bitBuffer2(std::vector<uint8_t>({0xAB, 0xF0}), bitSize);
    ASSERT_EQ(bitBuffer1.hashCode(), bitBuffer2.hashCode());

    const BitBuffer bitBuffer3(std::vector<uint8_t>({0xAB, 0xFF}), bitSize);
    ASSERT_EQ(bitBuffer1.hashCode(), bitBuffer3.hashCode());

    const BitBuffer bitBuffer4(std::vector<uint8_t>({0xAB, 0xC0}), bitSize);
    ASSERT_NE(bitBuffer1.hashCode(), bitBuffer4.hashCode());

    const BitBuffer bitBuffer5(std::vector<uint8_t>({0xBA, 0xE0}), bitSize);
    ASSERT_NE(bitBuffer1.hashCode(), bitBuffer5.hashCode());

    const BitBuffer bitBuffer6(std::vector<uint8_t>({0xAB}));
    ASSERT_NE(bitBuffer1.hashCode(), bitBuffer6.hashCode());

    const BitBuffer bitBuffer7;
    ASSERT_NE(bitBuffer1.hashCode(), bitBuffer7.hashCode());
}

TEST(BitBufferTest, constGetBuffer)
{
    const size_t bitSize = 11;
    const std::vector<uint8_t> buffer = {0xAB, 0xC0};
    const BitBuffer bitBuffer(buffer, bitSize);

    const uint8_t* readBuffer = bitBuffer.getBuffer();
    for (uint8_t element : buffer)
    {
        ASSERT_EQ(element, *readBuffer);
        readBuffer++;
    }
}

TEST(BitBufferTest, getBuffer)
{
    const size_t bitSize = 11;
    const std::vector<uint8_t> buffer = {0xAB, 0xC0};
    BitBuffer bitBuffer(buffer, bitSize);

    uint8_t* readBuffer = bitBuffer.getBuffer();
    for (uint8_t element : buffer)
    {
        ASSERT_EQ(element, *readBuffer);
        readBuffer++;
    }
}

TEST(BitBufferTest, getBitSize)
{
    const size_t bitSize = 11;
    const std::vector<uint8_t> buffer = {0xAB, 0xC0};
    const BitBuffer bitBuffer(buffer, bitSize);
    ASSERT_EQ(bitSize, bitBuffer.getBitSize());
}

TEST(BitBufferTest, getByteSize)
{
    const size_t bitSize = 11;
    const std::vector<uint8_t> buffer = {0xAB, 0xC0};
    const size_t byteSize = buffer.size();
    const BitBuffer bitBuffer(buffer, bitSize);
    ASSERT_EQ(byteSize, bitBuffer.getByteSize());
}

TEST(BitBufferTest, getData)
{
    const size_t bitSize = 11;
    const std::vector<uint8_t> buffer = {0xAB, 0xC0};
    const BitBuffer bitBuffer(buffer, bitSize);

    Span<const uint8_t> data = bitBuffer.getData();
    ASSERT_EQ(buffer.size(), data.size());
    for (size_t i = 0; i < buffer.size(); ++i)
    {
        ASSERT_EQ(buffer[i], data[i]);
    }
}

} // namespace zserio

File: compiler/extensions/java/runtime/test/zserio/runtime/io/BitBufferTest.java
package zserio.runtime.io;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.assertThrows;
import org.junit.jupiter.api.Test;

public class BitBufferTest
{
    @Test
    public void bufferConstructor()
    {
        final int byteSize = 2;
        final byte[] buffer = new byte[byteSize];
        final BitBuffer bitBuffer = new BitBuffer(buffer);
        assertEquals(8 * byteSize, bitBuffer.getBitSize());

        final long emptyBitSize = 0;
        final byte[] emptyBuffer = new byte[0];
        final BitBuffer emptyBitBuffer = new BitBuffer(emptyBuffer);
        assertEquals(emptyBitSize, emptyBitBuffer.getBitSize());
    }

    @Test
    public void bufferBitSizeConstructor()
    {
        final long bitSize = 11;
        final byte[] buffer = new byte[(int)((bitSize + 7) / 8)];
        final BitBuffer bitBuffer = new BitBuffer(buffer, bitSize);
        assertEquals(bitSize, bitBuffer.getBitSize());

        final long emptyBitSize = 0;
        final byte[] emptyBuffer = new byte[0];
        final BitBuffer emptyBitBuffer = new BitBuffer(emptyBuffer, emptyBitSize);
        assertEquals(emptyBitSize, emptyBitBuffer.getBitSize());

        final long outOfRangeBitSize = 9;
        final byte[] outOfRangeBuffer = new byte[1];
        assertThrows(IllegalArgumentException.class, () -> new BitBuffer(outOfRangeBuffer, outOfRangeBitSize));
    }

    @Test
    public void equalsMethod()
    {
        final long bitSize = 11;
        final BitBuffer bitBuffer1 = new BitBuffer(new byte[]{(byte)0xAB, (byte)0xE0}, bitSize);
        final BitBuffer bitBuffer2 = new BitBuffer(new byte[]{(byte)0xAB, (byte)0xF0}, bitSize);
        assertTrue(bitBuffer1.equals(bitBuffer2));

        final BitBuffer bitBuffer3 = new BitBuffer(new byte[]{(byte)0xAB, (byte)0xFF}, bitSize);
        assertTrue(bitBuffer1.equals(bitBuffer3));

        final BitBuffer bitBuffer4 = new BitBuffer(new byte[]{(byte)0xAB, (byte)0xC0}, bitSize);
        assertFalse(bitBuffer1.equals(bitBuffer4));

        final BitBuffer bitBuffer5 = new BitBuffer(new byte[]{(byte)0xBA, (byte)0xE0}, bitSize);
        assertFalse(bitBuffer1.equals(bitBuffer5));

        final BitBuffer bitBuffer6 = new BitBuffer(new byte[]{(byte)0xAB});
        assertFalse(bitBuffer1.equals(bitBuffer6));

        final BitBuffer bitBuffer7 = new BitBuffer(new byte[]{});
        assertFalse(bitBuffer1.equals(bitBuffer7));
    }

    @Test
    public void hashCodeMethod()
    {
        final long bitSize = 11;
        final BitBuffer bitBuffer1 = new BitBuffer(new byte[]{(byte)0xAB, (byte)0xE0}, bitSize);
        final BitBuffer bitBuffer2 = new BitBuffer(new byte[]{(byte)0xAB, (byte)0xF0}, bitSize);
        assertTrue(bitBuffer1.hashCode() == bitBuffer2.hashCode());

        final BitBuffer bitBuffer3 = new BitBuffer(new byte[]{(byte)0xAB, (byte)0xFF}, bitSize);
        assertTrue(bitBuffer1.hashCode() == bitBuffer3.hashCode());

        final BitBuffer bitBuffer4 = new BitBuffer(new byte[]{(byte)0xAB, (byte)0xC0}, bitSize);
        assertFalse(bitBuffer1.hashCode() == bitBuffer4.hashCode());

        final BitBuffer bitBuffer5 = new BitBuffer(new byte[]{(byte)0xBA, (byte)0xE0}, bitSize);
        assertFalse(bitBuffer1.hashCode() == bitBuffer5.hashCode());

        final BitBuffer bitBuffer6 = new BitBuffer(new byte[]{(byte)0xAB});
        assertFalse(bitBuffer1.hashCode() == bitBuffer6.hashCode());

        final BitBuffer bitBuffer7 = new BitBuffer(new byte[]{});
        assertFalse(bitBuffer1.hashCode() == bitBuffer7.hashCode());
    }

    @Test
    public void getBuffer()
    {
        final long bitSize = 11;
        final byte[] buffer = new byte[]{(byte)0xAB, (byte)0xE0};
        final BitBuffer bitBuffer = new BitBuffer(buffer, bitSize);
        assertTrue(java.util.Arrays.equals(buffer, bitBuffer.getBuffer()));
    }

    @Test
    public void getBitSize()
    {
        final long bitSize = 11;
        final byte[] buffer = new byte[]{(byte)0xAB, (byte)0xE0};
        final BitBuffer bitBuffer = new BitBuffer(buffer, bitSize);
        assertEquals(bitSize, bitBuffer.getBitSize());
    }

    @Test
    public void getByteSize()
    {
        final long bitSize = 11;
        final byte[] buffer = new byte[]{(byte)0xAB, (byte)0xE0};
        final int byteSize = buffer.length;
        final BitBuffer bitBuffer = new BitBuffer(buffer, bitSize);
        assertEquals(byteSize, bitBuffer.getByteSize());
    }
}
